api/types/events/api.go

```diff
@@ -77,11 +77,19 @@ type Emitter interface {
 	EmitAuditEvent(context.Context, AuditEvent) error
 }
 
+// PreparedSessionEvent is an event that has been prepared by
+// a [github.com/gravitational/teleport/lib/events.SessionEventSetter].
+// More specifically, it is a wrapper around an AuditEvent that signifies
+// the event has been prepared and is ready to be recorded or emitted.
+type PreparedSessionEvent interface {
```

Comment:
I wanted to avoid a exported concrete type so devs couldn't inadvertently create a prepared event without actually preparing it. The way it is now you have to use a preparer from `lib/events` or refactor the existing code so you can create a concrete prepared event yourself which would be a red flag.

Reply:
Yeah, I agree with that. 

Maybe this is one good use case for an exported method that returns an unexported type.

Reply:
I may be misremembering but I think our linter will whine about an exported method that returns an unexported type.

I think exporting a concrete type is fine, if a caller is willing to make the explicit cast, they are saying they know what they are doing, they can deal with the consequence. The interface kind of gets in the way, there's at least one place already where you create a nooppreparer just to convert the type, that's a perfect place to just do the type cast I think

Reply:
Just spitballing here, not even sure I like this idea...

You would have to move some things around - the code that creates prepared events would have to live in api/events instead of lib/events, but then you could have something like:

```
type preparedEvent struct { ... }

func PrepareEvent(event AuditEvent) preparedEvent { ... }

func Record(event preparedEvent) { ... }
```

Normally I'd push back against such a change, as it's kind of a pain to work with public APIs that require unexported types. In this case, it's kind of what you want though: there's only one way to get a `preparedEvent`, so you're forced into doing the right thing and calling `PrepareEvent` first.

If we export `PreparedEvent`, a new dev struggling to get their code to compile might just add an explicit type conversion and move on, not realizing that they broke the rules. (Though maybe Nic's right and this isn't a big deal)

This would also have impacts on your tests, because if you wanted to keep a no-op preparer, it would have to be in the api/events package.

What do you guys think? I'm still not sure, which probably means we shouldn't do it.

Reply:
Ah I understand what you mean now. I had the exact same idea, but you're unfortunately right in that it won't work inside `api/events`. Too much code would need `preparedEvent` inside `lib/events` which it wouldn't have access to. I tried moving `Stream` to `lib/events` so `preparedEvent` could live there as well, but that would require `api/events` importing `lib/events` which is a non-starter.